ereport/api/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! HTTP API for ereport producers.
+
+use dropshot::HttpError;
+use dropshot::HttpResponseOk;
+use dropshot::Path;
+use dropshot::Query;
+use dropshot::RequestContext;
+pub use ereport_types::Ena;
+pub use ereport_types::Ereport;
+pub use ereport_types::EreporterGenerationUuid;
+pub use ereport_types::Reporter;
+use openapi_manager_types::{
+    SupportedVersion, SupportedVersions, api_versions,
+};
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use std::num::NonZeroU32;
+use uuid::Uuid;
+
+api_versions!([
+    // WHEN CHANGING THE API (part 1 of 2):
+    //
+    // +- Pick a new semver and define it in the list below.  The list MUST
+    // |  remain sorted, which generally means that your version should go at
+    // |  the very top.
+    // |
+    // |  Duplicate this line, uncomment the *second* copy, update that copy for
+    // |  your new API version, and leave the first copy commented out as an
+    // |  example for the next person.
+    // v
+    // (next_int, IDENT),
+    (1, INITIAL),
+]);
+
+// WHEN CHANGING THE API (part 2 of 2):
+//
+// The call to `api_versions!` above defines constants of type
+// `semver::Version` that you can use in your Dropshot API definition to specify
+// the version when a particular endpoint was added or removed.  For example, if
+// you used:
+//
+//     (2, ADD_FOOBAR)
+//
+// Then you could use `VERSION_ADD_FOOBAR` as the version in which endpoints
+// were added or removed.
+
+/// API for ereport producers.
+#[dropshot::api_description]
+pub trait EreporterApi {
+    type Context;
+
+    /// Collect a tranche of ereports from this reporter.
+    #[endpoint {
+        method = GET,
```

Comment:
I do not have a strong feeling about this, but it sounds like this endpoint makes changes to the server (related to committed reports), and that's generally frowned upon in a `GET`.  Clients do often assume they can make `GET` requests without changing anything.  (The canonical example, which actually happened to me ages ago, is implementing a delete API using a link (which causes a browser to do `GET`s) and then having a search engine crawl all the delete links and delete everything.)  I don't know that we have automation today that does this, but it's not inconceivable.  It might be better to make this a `POST`.  I'd also document the behavior in the doc comment.

This behavior (whether `GET` or `POST`) is pretty non-RESTy, but I do see the advantage (fewer round-trips).

Reply:
This also felt icky to me; it's definitely not RESTful but none of this API is particularly RESTy...  Perhaps it's better off as a POST --- I'll note that (IIRC) gRPC just uses POST for _everything_ because it recognizes that some RPCs may not really fit precisely into any of the REST verbs. I _also_ find returning data in a POST a bit icky, but I think it's more correct to treat a side-effectful but ultimately idempotent operation as a POST than a GET; in particular, while this is retryable, it's definitely not cacheable.